File: prog/app/src/main/java/com/example/coincex/api_data/GlobalData.kt
package com.example.coincex.api_data

import android.content.Context
import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import org.json.JSONObject
import java.math.RoundingMode
import java.text.DecimalFormat

// Data class per le info globali del mercato
data class GlobalData(
    val marketCap: String,
    val volumeCap: String,
    val btcDom: String,
    val marketChange: String,
    val volumeChange: String,
    val btcDomChange: String) {

    // Oggetto contenente metodi per ottenere dati attraverso le API di CoinMarketCap
    companion object {

        // Metodo asincrono che restituisce un callback contenente il risultato della chiamata
        fun getDataFromApi(context: Context, callback:(result: String) -> Unit) {
            val url = "https://pro-api.coinmarketcap.com/v1/global-metrics/quotes/latest?CMC_PRO_API_KEY=e3dc6624-b531-4a8e-b501-7474e1e2455a"
            val queue = Volley.newRequestQueue(context)
            val stringRequest = StringRequest(
                Request.Method.GET, url,
                {
                    callback(it)
                }
            ) { callback("null") }
            queue.add(stringRequest)
        }

        // Metodo per il bind dei dati e restituisce la data class
        fun getData(data: String): GlobalData {

            val marketCap: String
            val volumeCap: String
            val btcDominance: String
            val marketChange: String
            val volumeChange: String
            val btcDominanceChange: String

            val df = DecimalFormat("#.##")
            df.roundingMode = RoundingMode.DOWN

            val jsonData = JSONObject(data).getJSONObject("data")
            marketCap = if (jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_market_cap") > 1000000000000)
                "$"+df.format(jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_market_cap")/1000000000000).toString()+" T"
            else
                "$"+df.format(jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_market_cap")/1000000000).toString()+" B"
            volumeCap = "$"+df.format(jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_volume_24h")/1000000000).toString()+" B"
            btcDominance = df.format(jsonData.getDouble("btc_dominance")).toString()+"%"
            marketChange = df.format(jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_market_cap_yesterday_percentage_change")).toString()+"%"
            volumeChange = df.format(jsonData.getJSONObject("quote").getJSONObject("USD").getDouble("total_volume_24h_yesterday_percentage_change")).toString()+"%"
            btcDominanceChange = df.format(jsonData.getDouble("btc_dominance_24h_percentage_change")).toString()+"%"

            return GlobalData(marketCap,volumeCap,btcDominance,marketChange,volumeChange,btcDominanceChange)
        }
    }
}

File: coincex_flutter/lib/src/ApiData/GlobalData.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class GlobalData {

  String marketCap;
  String volumeCap;
  String btcDom;
  String marketChange;
  String volumeChange;
  String btcDomChange;

  GlobalData(this.marketCap, this.volumeCap, this.btcDom, this.marketChange,
      this.volumeChange, this.btcDomChange);

  }

  Future<GlobalData> getGlobalData() async{
    var result = await http.get(Uri.parse("https://pro-api.coinmarketcap.com/v1/global-metrics/quotes/latest?CMC_PRO_API_KEY=e3dc6624-b531-4a8e-b501-7474e1e2455a"));;
    var toJson = jsonDecode(result.body);

    var marketCap = "";
    if (toJson["data"]["quote"]["USD"]["total_market_cap"] > 1000000000000) {
      marketCap = "${"\$"+(toJson["data"]["quote"]["USD"]["total_market_cap"]/1000000000000).toStringAsFixed(2)} T";
    }
    else {
      marketCap = "${"\$"+(toJson["data"]["quote"]["USD"]["total_market_cap"]/1000000000).toStringAsFixed(2)} B";
    }
    var volumeCap = "${"\$"+(toJson["data"]["quote"]["USD"]["total_volume_24h"]/1000000000).toStringAsFixed(2)} B";
    var btcDom = toJson["data"]["btc_dominance"].toStringAsFixed(2)+"%";

    var marketChange = "";
    if (toJson["data"]["quote"]["USD"]["total_market_cap_yesterday_percentage_change"] > 0) {
      marketChange = "${"+"+toJson["data"]["quote"]["USD"]["total_market_cap_yesterday_percentage_change"].toStringAsFixed(2)}%";
    }
    else {
      marketChange = toJson["data"]["quote"]["USD"]["total_market_cap_yesterday_percentage_change"].toStringAsFixed(2) + "%";
    }

    var volumeChange = "";
    if (toJson["data"]["quote"]["USD"]["total_volume_24h_yesterday_percentage_change"] > 0) {
      volumeChange = "${"+"+toJson["data"]["quote"]["USD"]["total_volume_24h_yesterday_percentage_change"].toStringAsFixed(2)}%";
    }
    else {
      volumeChange = toJson["data"]["quote"]["USD"]["total_volume_24h_yesterday_percentage_change"].toStringAsFixed(2)+"%";
    }

    var btcDomChange = "";
    if (toJson["data"]["btc_dominance_24h_percentage_change"] > 0) {
      btcDomChange = "${"+"+toJson["data"]["btc_dominance_24h_percentage_change"].toStringAsFixed(2)}%";
    }
    else {
      btcDomChange = toJson["data"]["btc_dominance_24h_percentage_change"].toStringAsFixed(2)+"%";
    }

    return GlobalData(marketCap, volumeCap, btcDom, marketChange, volumeChange, btcDomChange);
  }
